# Reanalysing data from the COVID-19 multiomics project

Full details of the original study can be found here: [https://dx.doi.org/10.1101%2F2020.07.17.20156513](https://dx.doi.org/10.1101%2F2020.07.17.20156513)

Authors provided a sql database with accession number `MSV000085703` from which multiomics data measurements were extracted. [ftp://massive.ucsd.edu/MSV000085703/other/Covid-19%20Study%20DB.sqlite](ftp://massive.ucsd.edu/MSV000085703/other/Covid-19%20Study%20DB.sqlite)

## Summary of data

*This information is taken from the original study by [Overmyer et al 2020](https://dx.doi.org/10.1101%2F2020.07.17.20156513)*

To gain biological insight into the host’s response to SARS-CoV-2 and pathways influencing its severity, we integrated our biomolecule measurements with clinical outcome variables. For this supervised analysis, we used univariate and multivariate regression to identify features that associate with (1) COVID-19 status and (2) HFD-45. Significant changes in plasma proteins, lipids, and metabolites associated with COVID-19 were determined by ANOVA and log-likelihood ratio tests, which incorporated potentially confounding variables, such as sex, age, and ICU status. Significant changes in leukocyte transcripts associated with COVID-19 diagnosis were analyzed separately by EB-seq. In total, 2,537 leukocyte transcripts, 146 plasma proteins, 168 plasma lipids, and 13 plasma metabolites were significantly associated with COVID-19 status (Table S1). 

Next, we applied GO and molecular class enrichment analysis to these features (Tables S2A and S2B). Among the biological processes that were enriched in biomolecules whose abundance was altered with COVID-19 (combined proteins and transcripts), many support recent findings (Messner et al., 2020) (Shen et al., 2020). These GO terms include mitotic cell cycle, phagocytosis recognition, positive recognition of B cell activation, complement activation (classical pathway), and innate immune response.

To find biomolecules associated with severity, we regressed HFD-45 against abundance of each biomolecule. Using this univariate approach that accounted for confounders sex and age, we found 6,202 transcripts, 189 plasma proteins, 218 plasma lipids, and 35 plasma small molecules that were associated with disease severity (Table S1). To further refine features of interest, for each molecule-type we performed multivariate linear regression on HFD-45 using the elastic net penalty (Zou and Hastie, 2005) (STAR Methods).

With the elastic net approach, 497 transcripts, 382 proteins, 140 lipids, and 60 metabolites were retained as predictive features for the outcome HFD-45 (Table S1). In total, we used the following criteria: (1) significance with COVID-19 status, (2) significance with HFD-45, and (3) elastic net feature selection, and generated a list of 219 features that were most significantly associated with COVID-19 status and severity (Figure 2C; Table S1).

In [1]:
import os
import pandas as pd
from functools import reduce

Tables in sqlite database:

```
biomolecules                   
deidentified_patient_metadata  
lipidomics_measurements        proteomics_measurements
lipidomics_runs                proteomics_runs
metabolomics_measurements      pvalues
metabolomics_runs              rawfiles
metadata                       transcriptomics_measurements
omes                           transcriptomics_runs
```

`.csv` files were extracted from the publicly accessible `sqlite` database. The following set of commands were repeated for each of the tables listed above.

```
.file 'Covid-19 Study DB.sqlite'
.header on
.mode csv

.output ${TABLE}.csv
select * from ${TABLE};
```

In [2]:
!for i in $(find . -name "*csv" | sort); do md5 $i; done

MD5 (./biomolecules.csv) = fe0f68f97ffdfef77c4a295cb57447a8
MD5 (./deidentified_patient_metadata.csv) = 14501da2f118dfc74cb26cfb91c6b8b4
MD5 (./lipidomics_measurements.csv) = c480651f7e570d37358bfe8ce5b76943
MD5 (./lipidomics_runs.csv) = 9816c081378a56b37c62bc7c3f1026d9
MD5 (./metabolomics_measurements.csv) = 16b8e7c19686816f9034f986835f0491
MD5 (./metabolomics_runs.csv) = a5301afb3f24bb43abdd044814d9e5a3
MD5 (./metadata.csv) = ffae3794ec2c21ec4513a6ca4028cbb1
MD5 (./omes.csv) = cdcb74634594b9991468c7ff5302e4c7
MD5 (./proteomics_measurements.csv) = 7cf5f63e2c30b692610ca1075446a159
MD5 (./proteomics_runs.csv) = 0f6e02c4277c5be3d05da71ef338d7e7
MD5 (./pvalues.csv) = db183bacd39ba6e15a1064890e554c55
MD5 (./rawfiles.csv) = 4aca8bd70db90e9a2fbe4d997188e057
MD5 (./transcriptomics_measurements.csv) = 005871ef93964adfecce2db5cce98a5c
MD5 (./transcriptomics_runs.csv) = e79131cf120d80b0175f0732ccfdf4a9


In [3]:
files = [x for x in os.listdir() if x.endswith("csv")]
data = [pd.read_csv(x, sep=",") for x in files]
data = dict(zip(files, data))
data.keys()

dict_keys(['proteomics_runs.csv', 'rawfiles.csv', 'proteomics_measurements.csv', 'lipidomics_measurements.csv', 'deidentified_patient_metadata.csv', 'transcriptomics_measurements.csv', 'transcriptomics_runs.csv', 'pvalues.csv', 'lipidomics_runs.csv', 'metabolomics_measurements.csv', 'metadata.csv', 'metabolomics_runs.csv', 'omes.csv', 'biomolecules.csv'])

Data is spread across individual `sql` tables. We remap the numerical codes to biologically meaningful identifiers. Database schema is available in Supp Fig S3 of the original manuscript.

In [4]:
def biomolecules_to_omes(data):
    gene = pd.DataFrame([[5, "Gene"]], columns=["omics_id", "omics_name"])
    omes = pd.concat([data["omes.csv"], gene])
    biomolecules = data["biomolecules.csv"]
    biomolecules = biomolecules.merge(omes, left_on="omics_id", right_on="omics_id", how="outer")
    return biomolecules

biomolecules = biomolecules_to_omes(data)
keep = biomolecules[biomolecules["keep"] == "1"]["biomolecule_id"].tolist()
biomolecules

biomolecule_id    standardized_name  omics_id keep  \
0                   1  unknown RT 4.081766         3    1   
1                   2  unknown RT 4.148561         3    1   
2                   3  unknown RT 5.261918         3    1   
3                   4  unknown RT 5.349937         3    1   
4                   5  unknown RT 5.462593         3    1   
...               ...                  ...       ...  ...   
27738           27795               ZYG11A         5    1   
27739           27796               ZYG11B         5    1   
27740           27797                  ZYX         5    1   
27741           27798                ZZEF1         5    1   
27742           27799                 ZZZ3         5    1   

                   omics_name  
0      Discovery Metabolomics  
1      Discovery Metabolomics  
2      Discovery Metabolomics  
3      Discovery Metabolomics  
4      Discovery Metabolomics  
...                       ...  
27738                    Gene  
27739                    Gene  
27740                    Gene  
27741                    Gene  
27742                    Gene  

[27743 rows x 5 columns]

## Selection of Outcome Measure

We intended to analyze the data with one encompassing outcome measure that fulfills the following criteria: 1) be able to combine severity of disease with mortality in one single metric; 2) be amenable to both ICU and medical floor populations; 3) use a timeframe that accounts for the fact that COVID-19 patients with respiratory failure require longer hospitalizations compared with non-COVID-19 individuals (Wang et al., 2020a, 2020b); 4) consider that COVID-19 causes a linear disease’s deterioration pattern that transition from mild respiratory compromise to respiratory failure, followed by respiratory distress requiring mechanical ventilatory support and eventually death. Thus, we selected the **composite outcome variable “hospital-free days at day 45” (HFD-45)**, which assigns zero value to patients requiring admission longer than 45 days or who die during the admission, and progressively more free days depending on the hospitalization length.

In [5]:
patient_metadata = data["deidentified_patient_metadata.csv"]
raw_files = data["rawfiles.csv"]
patient_files = raw_files.merge(patient_metadata, left_on="sample_id", right_on="sample_id", how="left")
patient_files.rename(columns={"keep": "keep_patient"}, inplace=True)
patient_files

rawfile_id     timestamp            rawfile_name  sample_id run_type  \
0             1  2.020043e+13  20200427_KAO_Control_1         -1  Control   
1             2  2.020043e+13  20200427_KAO_Control_2         -1  Control   
2             3  2.020043e+13  20200427_KAO_Control_3         -1  Control   
3             4  2.020043e+13  20200427_KAO_Control_4         -1  Control   
4             5  2.020043e+13   20200427_KAO_COVID_11         11   Sample   
..          ...           ...                     ...        ...      ...   
890         891 -1.000000e+00                   NC021        124   Sample   
891         892 -1.000000e+00                   NC022        125   Sample   
892         893 -1.000000e+00                   NC023        126   Sample   
893         894 -1.000000e+00                   NC024        127   Sample   
894         895 -1.000000e+00                   NC025        128   Sample   

     keep_patient  batch  ome_id Sample_label Albany_sampleID  ...  \
0               1      1       2          NaN             NaN  ...   
1               1      1       2          NaN             NaN  ...   
2               1      1       2          NaN             NaN  ...   
3               1      1       2          NaN             NaN  ...   
4               1      1       2     COVID_11             C11  ...   
..            ...    ...     ...          ...             ...  ...   
890             1      1       5  NONCOVID_21            NC21  ...   
891             1      1       5  NONCOVID_22            NC22  ...   
892             1      1       5  NONCOVID_23            NC23  ...   
893             1      1       5  NONCOVID_24            NC24  ...   
894             1      1       5  NONCOVID_25            NC25  ...   

     Vent_free_days   DM Hospital_free_days_45  Ferritin_NG/ML  CRP_MG/L  \
0               NaN  NaN                   NaN             NaN       NaN   
1               NaN  NaN                   NaN             NaN       NaN   
2               NaN  NaN                   NaN             NaN       NaN   
3               NaN  NaN                   NaN             NaN       NaN   
4              28.0  1.0                  37.0           401.0      64.1   
..              ...  ...                   ...             ...       ...   
890            28.0  1.0                  36.0           106.0       NaN   
891            28.0  0.0                  43.0             NaN      26.1   
892            28.0  0.0                  40.0            85.0      41.0   
893            28.0  0.0                  43.0             NaN       NaN   
894            28.0  0.0                  44.0             NaN       NaN   

     DDIMER_mg/L_FEU  Procalcitonin_NG/ML  Lactate_MMol/L  Fibrinogen  SOFA  
0                NaN                  NaN             NaN         NaN   NaN  
1                NaN                  NaN             NaN         NaN   NaN  
2                NaN                  NaN             NaN         NaN   NaN  
3                NaN                  NaN             NaN         NaN   NaN  
4                NaN                 0.21             NaN         NaN   NaN  
..               ...                  ...             ...         ...   ...  
890             1.46                  NaN            0.73       377.0   NaN  
891              NaN                 0.05             NaN         NaN   1.0  
892             1.87                 0.73            1.99       635.0   NaN  
893             0.19                  NaN            0.87         NaN   NaN  
894              NaN                  NaN            0.99         NaN   NaN  

[895 rows x 27 columns]

In [6]:
def biomolecules_to_data(biomolecules, data):
    # biomolecules = biomolecules[biomolecules["keep"] == "1"]
    return data.merge(biomolecules, left_on="biomolecule_id", right_on="biomolecule_id", how="left")

def measurements_to_runs(data, runs):
    return data.merge(runs, left_on="replicate_id", right_on="replicate_id", how="left")

def runs_to_patient(runs, patient):
    return runs.merge(patient, left_on="rawfile_id", right_on="rawfile_id", how="left")

omics = {"L": data['lipidomics_measurements.csv'],
         "M": data['metabolomics_measurements.csv'],
         "P": data['proteomics_measurements.csv'],
         "T": data['transcriptomics_measurements.csv'],}

runs = {"L": data['lipidomics_runs.csv'],
        "M": data['metabolomics_runs.csv'],
        "P": data['proteomics_runs.csv'],
        "T": data['transcriptomics_runs.csv'],}

omics = [measurements_to_runs(omics[i], runs[i]) for i in omics.keys()]
omics = dict(zip(runs.keys(), omics))
omics = [runs_to_patient(omics[i], patient_files) for i in omics.keys()]
omics = dict(zip(runs.keys(), omics))
data_omics = [biomolecules_to_data(biomolecules, omics[i]) for i in omics.keys()]
patient_status = "Albany_sampleID" # unique_identifier

## Lipidomics

The LC–MS data were processed using Compound Discoverer 2.1 (Thermo Scientific) and LipiDex (Hutchins et al., 2018) (v. 1.1.0). All peaks between 1 min and 45 min retention time and 100 Da to 5000 Da MS1 precursor mass were grouped into distinct chromatographic profiles (i.e., compound groups) and aligned using a 10-ppm mass and 0.3 min retention time tolerance. Profiles not reaching a minimum peak intensity of 5x10ˆ5, a maximum peak-width of 0.75, a signal-to-noise (S/N) ratio of 3, and a 3-fold intensity increase over blanks were excluded from further processing. MS/MS spectra were searched against an in-silico generated lipid spectral library containing 35,000 unique molecular compositions representing 48 distinct lipid classes (LipiDex library “LipiDex_HCD_Formic”, with a full range of acyl-chains included). Spectral matches with a dot product score greater than 500 and a reverse dot product score greater than 700 were retained for further analysis, with a minimum 75% spectral purity for designating fatty acid composition. Removed from the data set were adducts, class IDs greater than 3.5 median absolute retention time deviation (M.A.D. RT) of each other, and features found in less than 3 files. Data were additionally searched with Compound Discoverer 3.1 with the discovery metabolomics nodes for additional spectral matching to mzCloud and mzVault libraries but retaining the feature group and peak picking settings as detailed for the Compound Discoverer 2.1 analysis.

In [7]:
l = data_omics[0][["unique_identifier", "standardized_name", "normalized_abundance", "keep"]]
l = l[l["keep"] == "1"]
l.drop("keep", axis=1, inplace=True)
l = l.pivot(index="unique_identifier", columns="standardized_name")

# this maps the sample to patient
l_id = data_omics[0][["unique_identifier", patient_status]]
l_id.drop_duplicates(inplace=True)
l_id.set_index("unique_identifier", inplace=True)

l = l.merge(l_id, left_index=True, right_index=True, how="left")
l.reset_index(inplace=True)
l.set_index(["unique_identifier", patient_status], inplace=True)
l.sort_index(level=patient_status, inplace=True)

# remove null and assign sample groups
l_class = l.reset_index()
l_class.dropna(inplace=True)

l_class["covid_state"] = l_class[patient_status]
l_class["covid_state"].replace(regex=r'^C.*', value='Covid', inplace=True)
l_class["covid_state"].replace(regex=r'^NC.*', value='NonCovid', inplace=True)
l_map = l_class[["unique_identifier", patient_status, "covid_state"]]
# l_map.to_csv("lipidomics_class.tsv", sep="\t", index=False)

l_data = l_class.drop([patient_status, "covid_state"], axis=1)
l_data.set_index("unique_identifier", inplace=True)
l_data.columns = [y for x, y in l_data.columns]
# l_data.to_csv("lipidomics_data.tsv", sep="\t")
l_data

<ipython-input-7-e734fdf2417e>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  l_id.drop_duplicates(inplace=True)
/Users/tche0054/miniconda3/envs/mixomics/lib/python3.8/site-packages/pandas/core/reshape/merge.py:648: UserWarning: merging between different levels can give an unintended result (2 levels on the left,1 on the right)
  warnings.warn(msg, UserWarning)


AC 10:0_RT_6.936  AC 12:0_RT_7.955  \
unique_identifier                                                      
20200430_KAO_COVID_1                     18.137455         16.841960   
20200429_KAO_COVID_10                    20.481348         18.060484   
20200506_KAO_COVID_100                   22.325877         21.306323   
20200507_KAO_COVID_101                   20.561890         18.847766   
20200508_KAO_COVID_102                   22.285913         17.983300   
...                                            ...               ...   
20200427_KAO_NONCOVID_5                  18.841850         16.238782   
20200429_KAO_NONCOVID_6                  18.462152         17.197018   
20200429_KAO_NONCOVID_7                  19.626082         18.286611   
20200430_KAO_NONCOVID_8_reinject         18.437335         17.346805   
20200428_KAO_NONCOVID_9                  18.116135         16.858610   

                                  AC 13:0_RT_8.306  AC 14:0_RT_9.398  \
unique_identifier                                                      
20200430_KAO_COVID_1                     12.844352         14.300304   
20200429_KAO_COVID_10                    15.178616         15.901777   
20200506_KAO_COVID_100                   14.915154         19.313421   
20200507_KAO_COVID_101                   14.463795         17.707147   
20200508_KAO_COVID_102                   14.900186         14.223309   
...                                            ...               ...   
20200427_KAO_NONCOVID_5                  12.549287         14.224545   
20200429_KAO_NONCOVID_6                  12.948937         15.835207   
20200429_KAO_NONCOVID_7                  14.593291         16.611980   
20200430_KAO_NONCOVID_8_reinject         14.085722         15.122278   
20200428_KAO_NONCOVID_9                  13.083286         15.754600   

                                  AC 14:1_RT_8.361  AC 14:1_RT_8.51  \
unique_identifier                                                     
20200430_KAO_COVID_1                     15.468755        16.006314   
20200429_KAO_COVID_10                    17.327229        17.866290   
20200506_KAO_COVID_100                   19.789357        20.138474   
20200507_KAO_COVID_101                   17.832855        18.290478   
20200508_KAO_COVID_102                   14.349153        16.391772   
...                                            ...              ...   
20200427_KAO_NONCOVID_5                  15.288985        15.018646   
20200429_KAO_NONCOVID_6                  17.096651        17.129969   
20200429_KAO_NONCOVID_7                  18.448637        18.520732   
20200430_KAO_NONCOVID_8_reinject         16.690248        16.841577   
20200428_KAO_NONCOVID_9                  16.609426        16.645479   

                                  AC 16:0_RT_10.888  AC 16:0_RT_11.031  \
unique_identifier                                                        
20200430_KAO_COVID_1                      16.430707          16.996782   
20200429_KAO_COVID_10                     17.525513          18.879430   
20200506_KAO_COVID_100                    18.146506          19.411699   
20200507_KAO_COVID_101                    17.781891          18.991900   
20200508_KAO_COVID_102                    14.319578          15.952894   
...                                             ...                ...   
20200427_KAO_NONCOVID_5                   15.879065          16.379486   
20200429_KAO_NONCOVID_6                   17.125441          18.291986   
20200429_KAO_NONCOVID_7                   17.289492          18.614540   
20200430_KAO_NONCOVID_8_reinject          17.448357          18.386158   
20200428_KAO_NONCOVID_9                   17.238334          18.215268   

                                  AC 16:1_RT_9.841  AC 16:1_RT_9.93  ...  \
unique_identifier                                                    ...   
20200430_KAO_COVID_1                     14.918037        14.601523  ...   
20200429_KAO_COVID_10                    15.6

## Metabolomics

GC-MS raw files were processed using a software suite developed in-house that is available at https://github.com/coongroup. Following data acquisition, raw EI-GC/MS spectral data was deconvolved into chromatographic features and then grouped into features based on co-elution. Only features with at least 10 fragment ions and present in 33% of samples were kept. Feature groups from samples and background were compared, and only feature groups greater than 3-fold higher than background were retained. Compound identifications for the metabolites analyzed were assigned by comparing deconvolved high-resolution spectra against unit-resolution reference spectra present in the NIST 12 MS/EI library as well as to authentic standards run in-house. To calculate spectral similarity between experimental and reference spectra, a weighted dot product calculation was used. Metabolites lacking a confident identification were classified as “Unknown metabolites” and appended a unique identifier based on retention time. Peak heights of specified quant m/z were used to represent feature (metabolite) abundance. The data set was also processed through, where we applied a robust linear regression approach, rlm() function (Marazzi et al., 1993), non- log2 transformed intensity values versus run order, to normalize for run order effects on signal. AEX-LC-MS/MS: raw files were processed using Xcalibur Qual Browser (v4.0.27.10, Thermo Scientific) with results exported and further processed using Microsoft Excel 2010. The prepared standard solution was used to locate appropriate peaks for peak area analysis.

In [8]:
m = data_omics[1][["unique_identifier", "standardized_name", "normalized_abundance", "keep", "omics_id"]]
m = m[m["keep"] == "1"]
m.drop("keep", axis=1, inplace=True)

m_d = m[m["omics_id"] == 3]
m_d.drop("omics_id", axis=1, inplace=True)

m_t = m[m["omics_id"] == 4]
m_t.drop("omics_id", axis=1, inplace=True)

m_d = m_d.pivot(index="unique_identifier", columns="standardized_name")
m_d.columns = [x[1] for x in m_d.columns]

# this maps the sample to patient
m_id = data_omics[1][["unique_identifier", patient_status]]
m_id.drop_duplicates(inplace=True)
m_id.set_index("unique_identifier", inplace=True)

m_d = m_d.merge(m_id, left_index=True, right_index=True, how="left")
m_d.reset_index(inplace=True)
m_d.set_index(["unique_identifier", patient_status], inplace=True)
m_d.sort_index(level=patient_status, inplace=True)

# remove null and assign sample groups
m_d_class = m_d.reset_index()
m_d_class.dropna(inplace=True)

m_d_class["covid_state"] = m_d_class[patient_status]
m_d_class["covid_state"].replace(regex=r'^C.*', value='Covid', inplace=True)
m_d_class["covid_state"].replace(regex=r'^NC.*', value='NonCovid', inplace=True)

m_d_data = m_d_class.drop([patient_status, "covid_state"], axis=1)
m_d_data.set_index("unique_identifier", inplace=True)
keep = ~m_d_data.eq(m_d_data.iloc[:, 0], axis=0).all(1)
m_d_data = m_d_data[keep]
# m_d_data.to_csv("metabolomicsdiscovery_data.tsv", sep="\t")

m_d_map = m_d_class[["unique_identifier", patient_status, "covid_state"]]
m_d_map.set_index("unique_identifier", inplace=True)
m_d_map = m_d_map[keep].reset_index()
# m_d_map.to_csv("metabolomicsdiscovery_class.tsv", sep="\t", index=False)
m_d_data

/Users/tche0054/miniconda3/envs/mixomics/lib/python3.8/site-packages/pandas/core/frame.py:4305: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(
/Users/tche0054/miniconda3/envs/mixomics/lib/python3.8/site-packages/pandas/core/frame.py:4305: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(
<ipython-input-8-4d9ee4ad5520>:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  m_id.drop_duplicates(inplace=Tru

1,2-Propanediol, 2TMS derivative RT5.872253  \
unique_identifier                                                             
20200430_BP_COVID_1                                               22.525987   
20200429_BP_COVID_10                                              22.634601   
20200506_BP_COVID_100                                             22.129562   
20200507_BP_COVID_101                                             21.942203   
20200508_BP_COVID_102                                             21.875789   
...                                                                     ...   
20200427_BP_NONCOVID_5_25split                                    21.915568   
20200429_BP_NONCOVID_6                                            21.053214   
20200429_BP_NONCOVID_7                                            22.185698   
20200430_BP_NONCOVID_5                                            22.532433   
20200428_BP_NONCOVID_9_25split                                    22.578803   

                                2,3-Dihydroxybutanoic acid, tris(trimethylsilyl)- RT10.455827  \
unique_identifier                                                                               
20200430_BP_COVID_1                                                     13.898984               
20200429_BP_COVID_10                                                    17.851052               
20200506_BP_COVID_100                                                   13.340284               
20200507_BP_COVID_101                                                   17.441368               
20200508_BP_COVID_102                                                   17.880837               
...                                                                           ...               
20200427_BP_NONCOVID_5_25split                                          16.933270               
20200429_BP_NONCOVID_6                                                  15.214941               
20200429_BP_NONCOVID_7                                                  18.270518               
20200430_BP_NONCOVID_5                                                  15.350095               
20200428_BP_NONCOVID_9_25split                                          12.705430               

                                2-Hydroxybutyric acid, 2TMS derivative RT7.639387  \
unique_identifier                                                                   
20200430_BP_COVID_1                                                     22.089462   
20200429_BP_COVID_10                                                    20.819783   
20200506_BP_COVID_100                                                   21.038076   
20200507_BP_COVID_101                                                   22.640710   
20200508_BP_COVID_102                                                   22.706349   
...                                                                           ...   
20200427_BP_NONCOVID_5_25split                                          21.654459   
20200429_BP_NONCOVID_6                                                  21.659314   
20200429_BP_NONCOVID_7                                                  22.467372   
20200430_BP_NONCOVID_5                                                  23.122245   
20200428_BP_NONCOVID_9_25split                                          22.493095   

                                2-Ketoisocaproic acid mo-tms RT8.802217  \
unique_identifier                                                         
20200430_BP_COVID_1                                           18.241204   
20200429_BP_COVID_10                                          16.655539   
20200506_BP_COVID_100                                         16.948493   
20200507_BP_COVID_101                                         17.050310   
20200508_BP_COVID_102                                         17.796130   
...                                                                 ...   
20200427_BP_NONCOVID_5_25split      

In [9]:
m_t = m_t.pivot(index="unique_identifier", columns="standardized_name")
m_t.columns = [x[1] for x in m_t.columns]

m_t = m_t.merge(m_id, left_index=True, right_index=True, how="left")
m_t.reset_index(inplace=True)
m_t.set_index(["unique_identifier", patient_status], inplace=True)
m_t.sort_index(level=patient_status, inplace=True)

# remove null and assign sample groups
m_t_class = m_t.reset_index()
m_t_class.dropna(inplace=True)

m_t_class["covid_state"] = m_t_class[patient_status]
m_t_class["covid_state"].replace(regex=r'^C.*', value='Covid', inplace=True)
m_t_class["covid_state"].replace(regex=r'^NC.*', value='NonCovid', inplace=True)
m_t_map = m_t_class[["unique_identifier", patient_status, "covid_state"]]
# m_t_map.to_csv("metabolomicstargeted_class.tsv", sep="\t", index=False)

m_t_data = m_t_class.drop([patient_status, "covid_state"], axis=1)
m_t_data.set_index("unique_identifier", inplace=True)
# m_t_data.to_csv("metabolomicstargeted_data.tsv", sep="\t")
m_t_data

2,3-diP-glyceric Acid Total  \
unique_identifier                                          
20200430_Batch4_COVID-1                        15.929258   
20200429_Batch3_COVID-10                       14.709945   
20200506_Batch5_COVID-100                      14.431759   
20200507_Batch6_COVID-101                      16.747144   
20200508_Batch7_COVID-102                      14.330357   
...                                                  ...   
20200427_Batch1_NON-COVID-5                    16.666224   
20200429_Batch3_NON-COVID-6                    16.747144   
20200429_Batch3_NON-COVID-7                    15.457637   
20200430_Batch4_NON-COVID-8                    15.867858   
20200428_Batch2_NON-COVID-9                    13.632541   

                             2,3-dihydroxyisovaleric Acid  \
unique_identifier                                           
20200430_Batch4_COVID-1                         13.371777   
20200429_Batch3_COVID-10                        13.277578   
20200506_Batch5_COVID-100                       12.335949   
20200507_Batch6_COVID-101                       12.741888   
20200508_Batch7_COVID-102                       13.813781   
...                                                   ...   
20200427_Batch1_NON-COVID-5                     11.222795   
20200429_Batch3_NON-COVID-6                     12.857203   
20200429_Batch3_NON-COVID-7                     11.234817   
20200430_Batch4_NON-COVID-8                     12.401413   
20200428_Batch2_NON-COVID-9                     10.670656   

                             2-hydroxy-3-methylbutryic acid  \
unique_identifier                                             
20200430_Batch4_COVID-1                           20.310080   
20200429_Batch3_COVID-10                          18.860602   
20200506_Batch5_COVID-100                         19.351647   
20200507_Batch6_COVID-101                         20.654035   
20200508_Batch7_COVID-102                         18.769915   
...                                                     ...   
20200427_Batch1_NON-COVID-5                       20.253497   
20200429_Batch3_NON-COVID-6                       19.416995   
20200429_Batch3_NON-COVID-7                       19.208958   
20200430_Batch4_NON-COVID-8                       19.868159   
20200428_Batch2_NON-COVID-9                       20.385744   

                             2-methylcitric Acid  \
unique_identifier                                  
20200430_Batch4_COVID-1                16.010178   
20200429_Batch3_COVID-10               17.937328   
20200506_Batch5_COVID-100              15.444756   
20200507_Batch6_COVID-101              15.823765   
20200508_Batch7_COVID-102              16.908299   
...                                          ...   
20200427_Batch1_NON-COVID-5            14.726005   
20200429_Batch3_NON-COVID-6            16.580494   
20200429_Batch3_NON-COVID-7            14.344296   
20200430_Batch4_NON-COVID-8            15.680030   
20200428_Batch2_NON-COVID-9            15.649481   

                             2P Glyceric and 3P Glyceric Acids  \
unique_identifier                                                
20200430_Batch4_COVID-1                              16.153631   
20200429_Batch3_COVID-10                             17.441518   
20200506_Batch5_COVID-100                            16.680030   
20200507_Batch6_COVID-101                            17.296701   
20200508_Batch7_COVID-102                            15.151651   
...                                                        ...   
20200427_Batch1_NON-COVID-5                          17.241909   
20200429_Batch3_NON-COVID-6                          17.375175   
20200429_Batch3_NON-COVID-7                          16.666224   
20200430_Batch4_NON-COVID-8                          18.269565   
20200428_Batch2_NON-COVID-9                          16.376872   

                             3-Deoxy-2-keto-6P-gluconic Acid  \
unique_identifier                 

## Proteomics

Shotgun proteomics raw files were searched using MaxQuant quantitative software package (Cox et al., 2014) (version 1.6.10.43) against UniProt Homo Sapiens database (downloaded on 6.18.2019), containing protein isoforms and computationally predicted proteins. If not specified, default MaxQuant settings were used. LFQ quantification was performed using LFQ minimum ratio count of 1 and no MS/MS requirement for LFQ comparisons. iBAQ quantitation and “match between runs” were enabled with default settings. ITMS MS/MS tolerance was set to 0.35 Da. Lists of quantified protein groups were filtered to remove reverse identifications, potential contaminants, and proteins identified only by a modification site. LFQ abundance values were log2 transformed. Missing quantitative values were imputed by randomly drawing values from the left tail of the normal distribution of all measured protein abundance values (Tyanova et al., 2016). Protein groups that contained more than 50% missing values were removed from final analyses. Relative standard deviations (RSDs) for each protein group quantified across all seven technical replicates of healthy plasma controls were calculated, and proteins with RSD greater than 30% were removed from final analyses. PRM: identification and quantification of targeted peptides for PRM analysis were performed using Skyline open access software package (version 20.1). 4-5 most intense and specific transitions were used to quantify peptide abundances, and area-under-the-curve measurements for each peptide were exported for further analysis.

In [10]:
p = data_omics[2][["unique_identifier", "standardized_name", "normalized_abundance", "keep"]]
p = p[p["keep"] == "1"]
p.drop("keep", axis=1, inplace=True)
p = p.pivot(index="unique_identifier", columns="standardized_name")
p.columns = [x[1] for x in p.columns]

# this maps the sample to patient
p_id = data_omics[2][["unique_identifier", patient_status]]
p_id.drop_duplicates(inplace=True)
p_id.set_index("unique_identifier", inplace=True)

p = p.merge(p_id, left_index=True, right_index=True, how="left")
p.reset_index(inplace=True)
p.set_index(["unique_identifier", patient_status], inplace=True)
p.sort_index(level=patient_status, inplace=True)

# remove null and assign sample groups
p_class = p.reset_index()
p_class.dropna(inplace=True)

p_class["covid_state"] = p_class[patient_status]
p_class["covid_state"].replace(regex=r'^C.*', value='Covid', inplace=True)
p_class["covid_state"].replace(regex=r'^NC.*', value='NonCovid', inplace=True)
p_map = p_class[["unique_identifier", patient_status, "covid_state"]]
# p_map.to_csv("proteomics_class.tsv", sep="\t", index=False)

p_data = p_class.drop([patient_status, "covid_state"], axis=1)
p_data.set_index("unique_identifier", inplace=True)
# p_data.to_csv("proteomics_data.tsv", sep="\t")
p_data

<ipython-input-10-0c575882c4d7>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  p_id.drop_duplicates(inplace=True)


A0A024R6I7;A0A0G2JRN3  \
unique_identifier                                                           
20200509_ES_Albany_COVID19_batch4_single-shot_1                 37.995543   
20200507_ES_Albany_COVID19_batch3_single-shot_10                37.844902   
202005010_ES_Albany_COVID19_batch5_single-shot_100              37.787556   
202005011_ES_Albany_COVID19_batch6_single-shot_101              37.068980   
202005013_ES_Albany_COVID19_batch7_single-shot_102              38.004675   
...                                                                   ...   
20200514_ES_Albany_NON-COVID19_batch1_single-sh...              37.507418   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...              38.077710   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...              37.612697   
20200509_ES_Albany_NON-COVID19_batch4_single-sh...              37.368865   
20200504_ES_Albany_NON-COVID19_batch2_single-sh...              37.362348   

                                                    A0A075B6H9  A0A075B6I0  \
unique_identifier                                                            
20200509_ES_Albany_COVID19_batch4_single-shot_1      27.176361   28.294477   
20200507_ES_Albany_COVID19_batch3_single-shot_10     29.253091   29.108465   
202005010_ES_Albany_COVID19_batch5_single-shot_100   26.969061   28.103146   
202005011_ES_Albany_COVID19_batch6_single-shot_101   32.415410   28.430139   
202005013_ES_Albany_COVID19_batch7_single-shot_102   26.636994   28.346803   
...                                                        ...         ...   
20200514_ES_Albany_NON-COVID19_batch1_single-sh...   27.417478   27.946370   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...   28.885038   29.587350   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...   28.203312   29.300686   
20200509_ES_Albany_NON-COVID19_batch4_single-sh...   27.278525   27.394275   
20200504_ES_Albany_NON-COVID19_batch2_single-sh...   28.907273   27.497546   

                                                    A0A075B6I4;A0A1W2PQ80  \
unique_identifier                                                           
20200509_ES_Albany_COVID19_batch4_single-shot_1                 24.783368   
20200507_ES_Albany_COVID19_batch3_single-shot_10                27.163654   
202005010_ES_Albany_COVID19_batch5_single-shot_100              26.583481   
202005011_ES_Albany_COVID19_batch6_single-shot_101              26.307425   
202005013_ES_Albany_COVID19_batch7_single-shot_102              26.662073   
...                                                                   ...   
20200514_ES_Albany_NON-COVID19_batch1_single-sh...              20.876354   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...              28.359803   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...              25.932312   
20200509_ES_Albany_NON-COVID19_batch4_single-sh...              23.972724   
20200504_ES_Albany_NON-COVID19_batch2_single-sh...              26.019498   

                                                    A0A075B6I9  A0A075B6J1  \
unique_identifier                                                            
20200509_ES_Albany_COVID19_batch4_single-shot_1      23.926601   23.490571   
20200507_ES_Albany_COVID19_batch3_single-shot_10     26.153098   23.410928   
202005010_ES_Albany_COVID19_batch5_single-shot_100   25.303093   20.852103   
202005011_ES_Albany_COVID19_batch6_single-shot_101   26.740082   23.796498   
202005013_ES_Albany_COVID19_batch7_single-shot_102   25.585918   24.429053   
...                                                        ...         ...   
20200514_ES_Albany_NON-COVID19_batch1_single-sh...   23.634669   23.275261   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...   25.850146   25.118228   
20200507_ES_Albany_NON-COVID19_batch3_single-sh...   25.214008   24.135788   
20200509_ES_Albany_NON-COVID19_batch4_single-sh...   25.857805   23.682282   
20200504_ES_Albany_NON-COVID19_batch2_single-sh...   25.066775   23.864480   

                  

## RNA-Seq Data Processing

All RNA transcripts were downloaded from the NCBI refseq ftp site (wget ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/∗.rna.fna.gz ). Only mRNA (accessions NM_xxxx and XM_xxxx) and rRNA (excluding 5.8S) was then extracted, and immunoglobulin transcripts were downloaded from ENSEMBL (IG_C, IG_D, IG_J and IG_V ). We created a file mapping accession numbers to gene symbols, and then used rsem-prepare-reference to build a bowtie-2 reference database. Fastq files were trimmed and filtered using a custom algorithm tailored to improve quality scores and maximize retained reads in paired-end data. RNA-Seq expression estimation was performed by RSEM v 1.3.0 (parameters: seed-length=20, no-qualities, bowtie2-k=200, bowtie2-sensitivity-level=sensitive) (Li and Dewey, 2011), with bowtie-2 (v 2.3.4.1) for the alignment step (Langmead and Salzberg, 2012), using the custom hg38 reference described above. After the collation of expression estimates, hemoglobin transcripts were removed from further analysis, and TPM values were rescaled to total 1,000,000 in each sample. Differential Expression analysis was performed using the EBSeq package (v 1.26.0) (Leng et al., 2013) in R (v 3.6.2).

In [11]:
t = data_omics[3][["unique_identifier", "standardized_name", "normalized_abundance", "keep"]]
t = t[t["keep"] == "1"]
t.drop("keep", axis=1, inplace=True)
t = t.pivot(index="unique_identifier", columns="standardized_name")
t.columns = [x[1] for x in t.columns]

# this maps the sample to patient
t_id = data_omics[3][["unique_identifier", patient_status]]
t_id.drop_duplicates(inplace=True)
t_id.set_index("unique_identifier", inplace=True)

t = t.merge(t_id, left_index=True, right_index=True, how="left")
t.reset_index(inplace=True)
t.set_index(["unique_identifier", patient_status], inplace=True)

t.sort_index(level=patient_status, inplace=True)

# remove null and assign sample groups
t_class = t.reset_index()
t_class.dropna(inplace=True)

t_class["covid_state"] = t_class[patient_status]
t_class["covid_state"].replace(regex=r'^C.*', value='Covid', inplace=True)
t_class["covid_state"].replace(regex=r'^NC.*', value='NonCovid', inplace=True)
t_map = t_class[["unique_identifier", patient_status, "covid_state"]]
# t_map.to_csv("transcriptomics_class.tsv", sep="\t", index=False)

t_data = t_class.drop([patient_status, "covid_state"], axis=1)
t_data.set_index("unique_identifier", inplace=True)
# t_data.to_csv("transcriptomics_data.tsv", sep="\t")
t_data

<ipython-input-11-e241a1e1e4de>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  t_id.drop_duplicates(inplace=True)


A1BG  A2M  AAAS  AACS  AAGAB  AAK1  AAMDC  AAMP  AANAT  \
unique_identifier                                                           
C001                4.5  3.7   8.7   7.4    9.8   9.8    6.5  10.3    3.0   
C010                2.9  3.7   9.0   7.5    9.6   9.3    6.7  10.2    1.8   
C100                4.6  3.3   9.5   7.1   10.0   9.2    7.5  10.2    3.8   
C101                4.3  3.2   9.6   7.3    9.7   9.6    7.1  10.6    2.4   
C102                3.3  1.0   8.4   6.2   10.0  10.3    5.9   9.0    4.1   
...                 ...  ...   ...   ...    ...   ...    ...   ...    ...   
NC005               5.0  1.1   9.6   6.9    9.9   9.7    8.0  10.3    6.1   
NC006               6.0  3.9   9.9   7.7    9.6   9.5    7.3  11.1    1.5   
NC007               4.8  2.7   9.1   6.6    9.6   9.7    6.3  10.3    1.4   
NC008               5.3  2.4   9.1   7.2    9.4   9.7    6.6  10.4    2.8   
NC009               4.5  1.9   9.2   7.4    9.7   9.9    6.3  10.4    2.8   

                   AAR2  ...  ZWILCH  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  \
unique_identifier        ...                                                    
C001                9.3  ...     6.9    6.4   6.0   6.9   9.6     1.9     9.8   
C010                9.0  ...     7.7    8.7   4.3   5.8  10.0     3.5     9.4   
C100                9.3  ...     6.3    7.3   5.0   6.4   9.8     3.5     9.2   
C101                9.6  ...     7.0    7.0   5.4   7.6   9.6     3.9     8.7   
C102                8.6  ...     7.2    6.3   5.8   6.5   9.7     3.5     9.9   
...                 ...  ...     ...    ...   ...   ...   ...     ...     ...   
NC005               9.3  ...     6.1    5.6   4.7   6.5   9.6     4.6     9.2   
NC006              10.1  ...     5.9    6.6   5.7   6.5   8.9     2.1     8.1   
NC007               9.3  ...     6.8    5.3   5.3   7.0   9.9     2.6     9.4   
NC008               8.8  ...     5.5    5.4   4.6   6.2  10.0     3.7     9.6   
NC009               9.4  ...     7.2    6.3   6.3   7.9   9.3     2.5     9.0   

                    ZYX  ZZEF1  ZZZ3  
unique_identifier                     
C001               13.3   11.5   9.0  
C010               14.1   11.0   8.2  
C100               14.3   10.0   8.0  
C101               13.3   10.6   8.6  
C102               14.7   11.5   8.7  
...                 ...    ...   ...  
NC005              14.8   10.3   7.2  
NC006              13.7   10.3   7.9  
NC007              14.1   11.0   8.6  
NC008              15.3   11.0   8.0  
NC009              12.8   10.7   9.1  

[125 rows x 13263 columns]

In [12]:
# our framework only ingests data with matched samples
data_clean = [l_data, m_d_data, m_t_data, p_data, t_data]
maps = [l_map, m_d_map, m_t_map, p_map, t_map]
maps = [x.set_index("Albany_sampleID") for x in maps]
common = [set(x.index.tolist()) for x in maps]
common = set.intersection(*common)
maps = [x.loc[list(common)] for x in maps]
matched = [i.loc[j["unique_identifier"]] for i, j in tuple(zip(data_clean, maps))]
[x.shape for x in matched]

[(125, 3357), (125, 111), (125, 39), (125, 517), (125, 13263)]

In [13]:
# remap identifiers to sample types
def map_common_id(metadata, data):
    metadata = pd.DataFrame(metadata.set_index("unique_identifier")["Albany_sampleID"])
    data = pd.merge(metadata, data, left_index=True, right_index=True)
    data = data.reset_index().set_index("Albany_sampleID")
    data.index.name = None
    return data.drop("unique_identifier", axis=1)

# long feature names break the pipeline
def shorten_colnames(data):
    name_long = pd.Series(data.columns, name="long")
    data.columns = [x[:24] for x in data.columns]
    name_short = pd.Series(data.columns, name="short")
    name_map = pd.concat([name_short, name_long], axis=1)
    return data, name_map

classes_clean = [l_class, m_d_class, m_t_class, p_class, t_class]
data_final = [map_common_id(x, y) for x, y in tuple(zip(classes_clean, matched))]
data_short = [shorten_colnames(x) for x in data_final]
data_map = [x[1] for x in data_short]
data_final = [x[0] for x in data_short]

In [14]:
outdata = [
    "data_lipidomics.tsv" ,
    "data_metabolomicsdiscovery.tsv" ,
    "data_metabolomicstargeted.tsv" ,
    "data_proteomics.tsv" ,
    "data_transcriptomics.tsv" ,
]

[x.to_csv(y, sep="\t") for x, y in tuple(zip(data_final, outdata))]

[None, None, None, None, None]

In [15]:
outmeta = [
    "lipidomics_class.tsv" ,
    "metabolomicsdiscovery_class.tsv" ,
    "metabolomicstargeted_class.tsv" ,
    "proteomics_class.tsv" ,
    "transcriptomics_class.tsv" ,
]

def map_sample_type(metadata, data):
    data = pd.merge(metadata, data.set_index("Albany_sampleID"), left_index=True, right_index=True)
    data = data[~data.index.duplicated(keep='first')]
    return pd.DataFrame(data["covid_state"])

classes_final = [map_sample_type(x, y) for x, y in tuple(zip(data_final, classes_clean))]
[x.to_csv(y, sep="\t") for x, y in tuple(zip(classes_final, outmeta))]

[None, None, None, None, None]

In [16]:
outmap = [
    "lipidomics_featuremap.tsv" ,
    "metabolomicsdiscovery_featuremap.tsv" ,
    "metabolomicstargeted_featuremap.tsv" ,
    "proteomics_featuremap.tsv" ,
    "transcriptomics_featuremap.tsv" ,
]

[x.to_csv(y, sep="\t") for x, y in tuple(zip(data_map, outmap))]

[None, None, None, None, None]

In [17]:
# md5sum if you are on a linux machine, md5 for mac
!for i in $(find . -name "*tsv" | sort); do md5 $i; done

MD5 (./lipidomics_class.tsv) = 40362278a17cc9bb82a5d229441407fb
MD5 (./lipidomics_data.tsv) = 6e44e4084e8f5d6121fc37dedf683a5c
MD5 (./lipidomics_featuremap.tsv) = 3b66e71d980072ec5f805bc2718d25e2
MD5 (./metabolomicsdiscovery_class.tsv) = 40362278a17cc9bb82a5d229441407fb
MD5 (./metabolomicsdiscovery_data.tsv) = 85bf22252c39845fa9d3eb1a017f9567
MD5 (./metabolomicsdiscovery_featuremap.tsv) = 8f67dc9f0fa200370e876629fbe2f30a
MD5 (./metabolomicstargeted_class.tsv) = 40362278a17cc9bb82a5d229441407fb
MD5 (./metabolomicstargeted_data.tsv) = cade6ccd687045e9ef92a745a069269f
MD5 (./metabolomicstargeted_featuremap.tsv) = a0f861b12005788dda8d722033e0b8ec
MD5 (./proteomics_class.tsv) = 40362278a17cc9bb82a5d229441407fb
MD5 (./proteomics_data.tsv) = 5b01fb0ba896a40d1deec99e3ecb0c3e
MD5 (./proteomics_featuremap.tsv) = 55ac8687db048b9032a9599c0e7e9d30
MD5 (./transcriptomics_class.tsv) = 40362278a17cc9bb82a5d229441407fb
MD5 (./transcriptomics_data.tsv) = 932bf182d5f55cbc367ed5c751c13ab6
MD5 (./transcrip

In [18]:
# all these files are identical, choose one to copy to avoid confusion
!cp lipidomics_class.tsv classes_diablo.tsv

# move these files to staging location
!find . -name "*tsv" -exec mv {} ../../data/MSV000085703/{} \;